src/general/OptionsMenu.cs

```diff
@@ -379,142 +461,247 @@ private Viewport.MSAA MSAAIndexToResolution(int index)
         }
     }
 
+    private void CompareSettings()
+    {
+        // Enable the save and reset buttons if the current setting values differ from the saved ones.
+        if (Settings.Instance == savedSettings)
+        {
+            // Settings match
+            resetButton.Disabled = true;
+            saveButton.Disabled = true;
+        }
+        else
+        {
+            // Settings differ
+            resetButton.Disabled = false;
+            saveButton.Disabled = false;
+        }
+    }
+
     /*
-      GUI callbacks
+      GUI Control Callbacks
     */
 
     private void OnBackPressed()
     {
         GUICommon.Instance.PlayButtonPressSound();
 
-        // TODO: only save if something was changed
+        // If any settings have been changed, show a dialogue asking if the changes should be kept or
+        // discarded.
+        if (Settings.Instance != savedSettings)
+        {
+            backConfirmationBox.Visible = true;
+            return;
+        }
+
+        EmitSignal(nameof(OnOptionsClosed));
+    }
+
+    private void OnResetPressed()
+    {
+        GUICommon.Instance.PlayButtonPressSound();
 
-        // TODO: ask for saving the settings
+        // Restore and apply the old saved settings.
+        Settings.Instance.LoadFromObject(savedSettings);
         Settings.ApplyAll();
+        ApplySettingsToControls(Settings.Instance);
+
+        // Buttons should be disabled now.
+        resetButton.Disabled = true;
+        saveButton.Disabled = true;
+    }
+
+    private void OnSavePressed()
+    {
+        GUICommon.Instance.PlayButtonPressSound();
 
+        // Save the new settings to the config file.
         if (!Settings.Save())
         {
-            // TODO: show an error popup
-            GD.PrintErr("Couldn't save the settings");
+            GD.PrintErr("Failed to save new options menu settings.");
+            return;
         }
 
-        EmitSignal(nameof(OnOptionsClosed));
+        // Copy over the new saved settings.
+        savedSettings = Settings.Instance.Clone();
+
+        // Buttons should be disabled now.
+        resetButton.Disabled = true;
+        saveButton.Disabled = true;
     }
 
-    private void ResetConfirmSelected()
+    private void OnDefaultsPressed()
     {
-        Settings.ResetToDefaults();
+        GUICommon.Instance.PlayButtonPressSound();
+
+        defaultsConfirmationBox.Visible = true;
+    }
+
+    private void BackConfirmSelected()
+    {
+        backConfirmationBox.Visible = false;
```

Comment:
The dialog box will automatically hide if either buttons are pressed, so I think this is redundant